src/components/exercise.cjsx

```diff
@@ -2,166 +2,141 @@
 
-    if confirm('Are you sure you want to save?')
-      if ExerciseStore.isNew(@getId())
-        ExerciseStore.on 'created', @redirect
-        ExerciseActions.create(@getId(), ExerciseStore.get(@getId()))
-      else
-        ExerciseActions.save(@getId())
+  renderIntroTab: ->
+    id = @props.exerciseId
+    <BS.TabPane eventKey="intro" tab="Intro">
+      <div className="exercise-stimulus">
+        <label>Exercise Stimulus</label>
+        <textarea onChange={@updateStimulus} defaultValue={ExerciseStore.getStimulus(id)}></textarea>
+      </div>
+    </BS.TabPane>
 
-  publishExercise: ->
-    if confirm('Are you sure you want to publish?')
-      ExerciseActions.save(@getId())
-      ExerciseActions.publish(@getId())
+  renderMpqTabs: ->
+    {questions} = ExerciseStore.get(@props.exerciseId)
+    for question, i in questions
+      <BS.TabPane key={question.id} eventKey={"question-#{i}"} tab={"Question #{i+1}"}>
+        <Question id={question.id}
+          sync={@sync}
+          canMoveLeft={i isnt 0}
+          canMoveRight={i isnt questions.length - 1}
+          moveQuestion={@moveQuestion}
+          removeQuestion={_.partial(@removeQuestion, question.id)} />
+      </BS.TabPane>
 
-  renderLoading: ->
-    <div>Loading exercise: {@getId()}</div>
+  renderSingleQuestionTab: ->
+    {questions} = ExerciseStore.get(@props.exerciseId)
+    <BS.TabPane key={0} eventKey='question-0' tab='Question'>
+      <Question id={_.first(questions)?.id} sync={@sync} />
+    </BS.TabPane>
 
-  renderFailed: ->
-    <div>Failed loading exercise, please check id</div>
+  previewTag: (tag) ->
+    content = _.compact([tag.name, tag.description]).join(' ') or tag.id
+    isLO = _.include(['lo', 'aplo'], tag.type)
+    {content, isLO}
 
-  sync: -> ExerciseActions.sync(@getId())
 
-  create: (id) ->
-    template = ExerciseStore.getTemplate()
-    ExerciseActions.loaded(template, id)
+  exercisePreviewData: (ex) ->
+    content: ex
+    tags: _.map ex.tags, (tag) -> name: tag
 
-  renderForm: ->
-    id = @getId()
+  addQuestion: ->
+    ExerciseActions.addQuestionPart(@props.exerciseId)
 
-    questions = []
-    for question in ExerciseStore.getQuestions(id)
-      questions.push(<Question key={question.id} sync={@sync} id={question.id} />)
+  selectTab: (tab) -> @setState({tab})
 
-    isWorking = ExerciseStore.isSaving(id) or ExerciseStore.isPublishing(id)
+  getActiveTab: (showMPQ) ->
+    if (not @state.tab or @state.tab?.indexOf('question-') is -1)
+      return @state.tab
 
-    if not ExerciseStore.isPublished(id) and not ExerciseStore.isNew(id)
-      publishButton = <AsyncButton
-        bsStyle='primary'
-        onClick={@publishExercise}
-        disabled={isWorking}
-        isWaiting={ExerciseStore.isPublishing(id)}
-        waitingText='Publishing...'
-        isFailed={ExerciseStore.isFailed(id)}
-        >
-        Publish
-      </AsyncButton>
+    question = @state.tab.split('-')[1]
```

Comment:
hmm, is there no better way to find the question index than parsing `"question-42"`?